src/ERC7739.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {MessageHashUtils} from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import {ERC7739Utils} from "./libraries/ERC7739Utils.sol";
+import {EIP712} from "./EIP712.sol";
+import {Key, KeyLib} from "./libraries/KeyLib.sol";
+import {TypedDataSignLib} from "./libraries/TypedDataSignLib.sol";
+import {PersonalSignLib} from "./libraries/PersonalSignLib.sol";
+
+/// @title ERC7739
+/// @notice An abstract contract that implements the ERC-7739 standard
+/// @notice This contract assumes that all data verified through ERC-1271 `isValidSignature` implements the defensive nested hashing scheme defined in EIP-7739
+/// @dev See https://eips.ethereum.org/EIPS/eip-7739
+abstract contract ERC7739 is EIP712 {
```

Comment:
High level comment.. but why is this a contract? It seems like it might just be better as a re-useable library? All of the functions defined here are private and I feel like we could re-architect it in a way where we wouldnt have to inherit from 712 and rather pass in any domain separator info

Reply:
Ok I moved some funcs to the library so this base contract is mostly just `_isValidTypedDataSig` and `_isValidNestedPersonalSignature`. I think it might be a bit tricky to not inherit from 712 because we need to use the 712 domain in two different ways (domain separator for personal sign) and domainBytes for nested 712

Reply:
I guess the only thing I don't like about this structure is that EIP712 is something we need to support regardless of 7739...

If one day we decide 7739 is no longer needed, they are very tightly coupled in our contracts right now

Reply:
Can we change the inheritance so that 712 is not a child of 7739?

Reply:
Ok yeah I took it out